custom_components/localtuya/core/pytuya/__init__.py

```diff
@@ -811,7 +811,9 @@ def __init__(
         self.remote_nonce = b""
         self.dps_whitelist = UPDATE_DPS_WHITELIST
         self.dispatched_dps = {}  # Store payload so we can trigger an event in HA.
-        self._last_command_sent = 1
+        self._last_command_sent = 1        # The time last command was sent
+        self._last_command_send_seqno = 0  # The number in the queue to send
```

Comment:
can you give an example of what and how would cause/fix this? because I was gonna refactor this soon and get rid of _last_command_sent, and use async methods better.

Reply:
First of all:
> I was gonna refactor this soon

With all the exising commits in the master, I'd suggest to make a new release first, and only then continue refactoring.

Now to this commit:

Each `await` allows other tasks in the loop queue to run, until they either finished, or entered their `await`s. It means, while one `transport_write` waits to ensure a delay between writes, other `transport_write` calls may jump in and do their writes, delaying the execution of the first one further and further. The `wait` counter I introduced a long ago was a dirty fix, because I found no a better solution that day. The counter may lead to writes with smaller delay and then the gateway to drop the connection: I saw in my logs such events!

But the more often problem I experienced is out of order commands sent to a particular device. I have chandeliers with a mix of Zigbee and BLE bulbs. Each chandelier is used in one of the 3 modes: bright white, scene, night (dim white light of one bulb only). Automations to turn a chandelier on into a particular mode, or to switch between modes, require several commands to send to the gateway. Moreover, each bulb requires more than one command to be executed in a particualr order! And sometimes the order of commands breaks, which is visible.

For example, Zigbee bulbs don't like to switch from "scene" to "white" being turned off: they enter "color" mode instead. So, if a chandelier was turned off from the scene mode, its Zigbee bulb should first be turned on, then be switched to white mode. But sometimes they are turned on in a random color, the color in scene mode at which the bulb was turned off! Exactly because the commands were executed in different order.

This commit ensures that all the writes are performed in the order of `transport_write` calls, and the delays between writes is never less than 0.05 seconds. So, it is _the_ solution.

Reply:
> What if we used asyncio.lock() instead

You are always welcome with your knowledge of Python 😄 But `wait` is not required here anymore as well. I'll implement and test it: I know how to generate bulk write requests.